### PR TITLE
[Doc] Update Data Preparation and Visualization doc

### PR DESCRIPTION
**Pull request by MengzhangLI**

### CODE REVIEW

File: docs/en/user_guides/visualization.md [hunk 1]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
 tensorboard --logdir work_dirs/test_visual/20220810_115248/vis_data
 ```
 
-If you would like to know more visualization usage, you can refer to [visualization tutorial](https://mmengine.readthedocs.io/en/latest/advanced_tutorials/visualization.html) in mmengie.
+We could use `SegLocalVisualizer` to visualize results with their corresponding predictions or ground truth. For example, please follow commands below in your MMSegmentation main directory:
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:79]: Add a subtitle
```
### Visualize a Single Data Sample
```

File: docs/en/user_guides/visualization.md [hunk 2]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
...
+from mmseg.structures import SegDataSample
+from mmseg.visualization import SegLocalVisualizer
+
+out_file = 'out_file_cityscapes'
+save_dir = './work_dirs'
+
+image = mmcv.imread(
+    osp.join(
+        osp.dirname(__file__),
+        './tests/data/pseudo_cityscapes_dataset/leftImg8bit/val/frankfurt/frankfurt_000000_000294_leftImg8bit.png'  # noqa
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:96]: Pseudo data is not recommended and proper in examples.

I suggest tell user use local dataset path.
just like https://github.com/open-mmlab/mmengine/blob/main/docs/zh_cn/tutorials/config.md

File: docs/en/user_guides/visualization.md [hunk 3]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
...
+    ),
+    'color')
+sem_seg = mmcv.imread(
+    osp.join(
+        osp.dirname(__file__),
+        './tests/data/pseudo_cityscapes_dataset/gtFine/val/frankfurt/frankfurt_000000_000294_gtFine_labelTrainIds.png'  # noqa
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:102]: same problem as above

File: docs/en/user_guides/visualization.md (hunk 1 above)
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:79]: add a subtitle before line 49
```
### Visualizer Data Samples during Model Testing or Validataion
```

Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:79]: I think this line is not proper as below is no any result only ground truth, so I suggest to change as
```
If you want to visualize a single data sample, a pair of input image and its ground truth in the dataset, you can use the scrips below to see them.
```

File: docs/en/user_guides/visualization.md [hunk 4]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
...
-If you would like to know more visualization usage, you can refer to [visualization tutorial](https://mmengine.readthedocs.io/en/latest/advanced_tutorials/visualization.html) in mmengie.
+We could use `SegLocalVisualizer` to visualize results with their corresponding predictions or ground truth. For example, please follow commands below in your MMSegmentation main directory:
+
+```python
+import mmcv
+import os.path as osp
+import torch
+from mmengine.structures import PixelData
+
+from mmseg.structures import SegDataSample
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:87]: Please give some explanation or a link to introduce `SegDataSample` or `PixelData`

File: docs/en/user_guides/visualization.md [hunk 5]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
...
+        './tests/data/pseudo_cityscapes_dataset/leftImg8bit/val/frankfurt/frankfurt_000000_000294_leftImg8bit.png'  # noqa
+    ),
+    'color')
+sem_seg = mmcv.imread(
+    osp.join(
+        osp.dirname(__file__),
+        './tests/data/pseudo_cityscapes_dataset/gtFine/val/frankfurt/frankfurt_000000_000294_gtFine_labelTrainIds.png'  # noqa
+    ),
+    'unchanged')
+sem_seg = torch.unsqueeze(torch.from_numpy(sem_seg), 0)
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:105]: It might not necessary to extend the dimension, as it allows input 2d data.
or if want to keep this line, I suggest add some comments, as it might make users confused about the dimension extension

https://github.com/open-mmlab/mmengine/blob/6b1b8a375163ea1b8712cf886204a1d08d17cee1/mmengine/structures/pixel_data.py#L79-L81

File: docs/en/user_guides/visualization.md [hunk 6]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
...
+             'vegetation', 'terrain', 'sky', 'person', 'rider',
+             'car', 'truck', 'bus', 'train', 'motorcycle',
+             'bicycle'),
+    palette=[[128, 64, 128], [244, 35, 232], [70, 70, 70],
+             [102, 102, 156], [190, 153, 153], [153, 153, 153],
+             [250, 170, 30], [220, 220, 0], [107, 142, 35],
+             [152, 251, 152], [70, 130, 180], [220, 20, 60],
+             [255, 0, 0], [0, 0, 142], [0, 0, 70],
+             [0, 60, 100], [0, 80, 100], [0, 0, 230],
+             [119, 11, 32]])
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:127]: add some comments about dataset meta information where users can find them.

File: docs/en/user_guides/visualization.md [hunk 7]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
...
+seg_local_visualizer.add_datasample(out_file, image,
+                                    data_sample)
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:129]: please introduce `show` option.

File: docs/en/user_guides/visualization.md [hunk 8]
```diff
@@ -76,4 +76,59 @@ we can also run the following command to view them in TensorBoard:
...
+```
+
+Then the visualization result of image with its corresponding ground truth could be found in `./work_dirs/vis_data/vis_image/` whose name is `out_file_cityscapes_0.png`.
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:132]: please add an image.

File: docs/en/user_guides/visualization.md [hunk 9]
```diff
@@ -76,4 +76,89 @@ we can also run the following command to view them in TensorBoard:
...
+data_sample = SegDataSample()
+data_sample.gt_sem_seg = gt_sem_seg
+
+seg_local_visualizer = SegLocalVisualizer(
+    vis_backends=[dict(type='LocalVisBackend')],
+    save_dir=save_dir)
+
+# The meta information of dataset usually includes `classes` for class names and
+# `palette` for visualization color of each foreground.
+# It is usually defined in corresponding class of dataset such as './mmseg/datasets/cityscapes.py'
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:137]: I suggest use this file https://github.com/open-mmlab/mmsegmentation/blob/1.x/mmseg/utils/class_names.py

All class names and palettes are defined in the file ...

File: docs/en/user_guides/visualization.md [hunk 10]
```diff
@@ -76,4 +76,89 @@ we can also run the following command to view them in TensorBoard:
...
+wget https://user-images.githubusercontent.com/24582831/189833143-15f60f8a-4d1e-4cbb-a6e7-5e2233869fac.png --output-document aachen_000000_000019_gtFine_labelTrainIds.png
+```
+
+Then you can find their local path and use the scrips below to visualize:
+
+```python
+import mmcv
+import os.path as osp
+import torch
+# `PixelData` is data structure for pixel-level annotations or predictions defined in MMEngine.
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:101]:
```suggestion
# `PixelData` is data structure for pixel-level annotations or predictions defined in MMEngine.
# Please refer to this tutorial  file URL of data structures in MMEngine 
```

File: docs/en/user_guides/visualization.md [hunk 11]
```diff
@@ -76,4 +76,89 @@ we can also run the following command to view them in TensorBoard:
...
+from mmengine.structures import PixelData
+
+# `SegDataSample` is data structure interface between different components
+# defined in MMSegmentation, it includes ground truth, prediction and
+# predicted logits of semantic segmentation.
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:106]:
```suggestion
# `SegDataSample` is data structure defined in MMSegmentation, 
# it includes ground truth, prediction and
# predicted logits of semantic segmentation.
# Please refer to this tutorial fileurl of `SegDataSample` for more details
```

File: docs/en/user_guides/2_dataset_prepare.md [hunk 12]
```diff
@@ -145,7 +145,7 @@ mmsegmentation
 The data could be found [here](https://www.cityscapes-dataset.com/downloads/) after registration.
 
 By convention, `**labelTrainIds.png` are used for cityscapes training.
-We provided a [scripts](https://github.com/open-mmlab/mmsegmentation/blob/master/tools/dataset_converters/cityscapes.py) based on [cityscapesscripts](https://github.com/mcordts/cityscapesScripts)
+We provided a [scripts](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/tools/dataset_converters/cityscapes.py) based on [cityscapesscripts](https://github.com/mcordts/cityscapesScripts)
```
Comment (MeowZheng) [inline docs/en/user_guides/2_dataset_prepare.md:148]: 1.x branch

File: docs/en/user_guides/visualization.md [hunk 13]
```diff
@@ -46,7 +46,9 @@ tensorboard --logdir work_dirs/test_visual/20220810_115248/vis_data
 
 ## Data and Results visualization
 
-MMSegmentation provides `SegVisualizationHook` that can render segmentation masks of ground truth and prediction. Users can modify `default_hooks` at each `schedule_x.py` config file.
+### Visualizer Data Samples during Model Testing or Validation
+
+MMSegmentation provides `SegVisualizationHook` that can visualize segmentation masks of ground truth and prediction in validation and testing process. It is defined in [here](https://github.com/open-mmlab/mmsegmentation/blob/1.x/mmseg/engine/hooks/visualization_hook.py#L16-L100) and users can modify `default_hooks` at each `schedule_x.py` config file.
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:51]: I suggest user guides do not cover the implementation of functions  or classes.

```suggestion
MMSegmentation provides `SegVisualizationHook` which is a [hook](URL for hook doc in engine) working to visualize ground truth and prediction of segmentation during model testing and evaluation. Its configuration is in `defualt_hooks`, please see [Runner's tutorial](URL for runner docs in engine) for more details.
```

File: docs/en/user_guides/visualization.md [hunk 14]
```diff
@@ -76,4 +78,94 @@ we can also run the following command to view them in TensorBoard:
 tensorboard --logdir work_dirs/test_visual/20220810_115248/vis_data
 ```
 
-If you would like to know more visualization usage, you can refer to [visualization tutorial](https://mmengine.readthedocs.io/en/latest/advanced_tutorials/visualization.html) in mmengie.
+### Visualize a Single Data Sample
+
+If you want to visualize a single data sample, a pair of input image and its ground truth in the dataset are necessary.
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:83]: ' are necessary' is grammatical error

Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:83]: 'If' is just a phrase, not a full sentence
```suggestion
If you want to visualize a single data sample, a pair of input image and its ground truth in the dataset, we suggest to use `SegLocalVisualizer`. Here is an example about `SegLocalVisualizer`, please download the example data first.
```

File: docs/en/user_guides/visualization.md [hunk 15]
```diff
@@ -76,4 +78,94 @@ we can also run the following command to view them in TensorBoard:
...
+<div align=center>
+<img src="https://user-images.githubusercontent.com/24582831/189833109-eddad58f-f777-4fc0-b98a-6bd429143b06.png" width="70%"/>
+</div>
+
+```shell
+wget https://user-images.githubusercontent.com/24582831/189833109-eddad58f-f777-4fc0-b98a-6bd429143b06.png --output-document aachen_000000_000019_leftImg8bit.png
+wget https://user-images.githubusercontent.com/24582831/189833143-15f60f8a-4d1e-4cbb-a6e7-5e2233869fac.png --output-document aachen_000000_000019_gtFine_labelTrainIds.png
+```
+
+Then you can find their local path and use the scrips below to visualize:
```
Comment (MeowZheng) [inline docs/en/user_guides/visualization.md:95]:
```suggestion
`SegLocalVisualizer` is a derivative from `Visualizer` in MMEngine and works for balabala. ...., for more details about `Visualizer` please refer to this tutorial.
```

Comment (MengzhangLI) [reply docs/en/user_guides/visualization.md:95]: I removed this comments in line 83 where `SegLocalVisualizer` is introduced.
